### PR TITLE
`MockHTTPClient`: encode the entire `HTTPRequest` into snapshots

### PR DESCRIPTION
**Pull request by NachoSoto**
This builds up on top of all the prior PRs that allow for easily encoding requests into simple `struct`s.
Prior to this, tests had to manually check the different parameters of requests (path, headers, body, method, etc.). Now it's all automatic and encoded together in the snapshot files.

Example snapshot:
```json
{
  "headers" : {
    "Authorization" : "Bearer asharedsecret"
  },
  "request" : {
    "body" : {
      "fetch_token" : "AQID",
      "product_identifiers" : [
        "producta",
        "productb",
        "productc",
        "productd"
      ]
    },
    "method" : "POST",
    "url" : "https://api.revenuecat.com/v1/subscribers/user/intro_eligibility"
  }
}
```

### CODE REVIEW

File: PurchasesTests/Mocks/MockHTTPClient.swift [hunk 1]
```diff
@@ -71,38 +72,42 @@ class MockHTTPClient: HTTPClient {
...
+        case url
+
+    }
 
-    // fixme: use SnapshotTesting to compare the whole `HTTPRequest` instead of only `requestBody`.
-    func expectToEqual(_ other: MockHTTPClient.Call, file: FileString = #file, line: UInt = #line) throws {
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
 
-        // Body comparison is done by SnapshotTesting
```
Comment (NachoSoto) [inline PurchasesTests/Mocks/MockHTTPClient.swift:79]: This was incomplete anyway, now unnecessary!

File: PurchasesTests/TestHelpers/SnapshotTesting+Extensions.swift [hunk 2]
```diff
@@ -37,12 +37,22 @@ private extension Encodable {
...
 
     func asFormattedData() throws -> Data {
+        var formatting: JSONEncoder.OutputFormatting {
+            var result: JSONEncoder.OutputFormatting = [
+                .prettyPrinted,
+                .sortedKeys
+            ]
+
+            if #available(iOS 13.0, tvOS 13.0, *) {
+                result.insert(.withoutEscapingSlashes)
```
Comment (aboedo) [inline PurchasesTests/TestHelpers/SnapshotTesting+Extensions.swift:47]: does this mean we get a different result when running on iOS 12 devices?

**Review by aboedo — APPROVED** [review]

File: PurchasesTests/TestHelpers/SnapshotTesting+Extensions.swift (hunk 2 above)
Comment (NachoSoto) [reply PurchasesTests/TestHelpers/SnapshotTesting+Extensions.swift:47]: Oh yeah, I think this is gonna fail on `main` when we run tests on iOS 12.
I guess we can remove this. It just means that URLs will be encoded in JSON like this: `"https:\/\/api.revenuecat.com\/v1\/subscribers\/user\/intro_eligibility"`

Comment (aboedo) [reply PurchasesTests/TestHelpers/SnapshotTesting+Extensions.swift:47]: agreed, I think it's better to have backwards compatibility 👍

**Comment by taquitos** [comment]
This PR makes me very happy.

**Pull request merged**